File: app/src/main/java/com/shen/somebase/widget/manager_drag_adapter/BaseItemTouchHelpCallBack.java
package com.shen.somebase.widget.manager_drag_adapter;

import androidx.annotation.Nullable;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.RecyclerView;

/*
 * created by shen at 2019/10/11 16:03
 */
public class BaseItemTouchHelpCallBack<T extends RecyclerView.Adapter> extends ItemTouchHelper.Callback {

    //是否可以长按拖住啊哦
    private boolean isLongDragEnable;

    //是否可以左右滑动
    private boolean isSwipeEnable;

    @Nullable
    private OnItemTouchCallbackListener<T> listener = null;

    @Nullable
    private ItemStatusChange listenerStatus = null;

    BaseItemTouchHelpCallBack(ITouchHelper touchHelper) {
        this.isLongDragEnable = touchHelper.isLongDragEnable();
        this.isSwipeEnable = touchHelper.isSwipeEnable();
        this.listener = touchHelper.listener();
        this.listenerStatus = touchHelper.listenerStatus();
    }

    @Override
    public boolean isLongPressDragEnabled() {
        return isLongDragEnable;
    }

    @Override
    public boolean isItemViewSwipeEnabled() {
        return isSwipeEnable;
    }

    @Override
    public int getMovementFlags(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder) {
        if (recyclerView.getLayoutManager() instanceof GridLayoutManager) {
            final int dragFlags = ItemTouchHelper.UP | ItemTouchHelper.DOWN | ItemTouchHelper.LEFT | ItemTouchHelper.RIGHT;
            final int swipeFlags = 0;
            return makeMovementFlags(dragFlags,swipeFlags);
        } else {
            final int dragFlags = ItemTouchHelper.UP | ItemTouchHelper.DOWN;
            final int swipeFlags = 0;
            return makeMovementFlags(dragFlags,swipeFlags);
        }
    }

    @Override
    public boolean onMove(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder, RecyclerView.ViewHolder target) {
        if (listener != null) {
            return listener.onMove(viewHolder.getAdapterPosition(), target.getAdapterPosition(), (T) recyclerView.getAdapter());
        }
        return false;
    }

    @Override
    public void onSwiped(RecyclerView.ViewHolder viewHolder, int direction) {
        if (listener != null) {
            listener.onSwiped(viewHolder.getAdapterPosition());
        }
    }

    @Override
    public void onSelectedChanged(RecyclerView.ViewHolder viewHolder, int actionState) {
        super.onSelectedChanged(viewHolder, actionState);
        if (listenerStatus != null) listenerStatus.onSelectChange(viewHolder, actionState);
    }

    @Override
    public void clearView(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder) {
        super.clearView(recyclerView, viewHolder);
        if (listenerStatus != null) listenerStatus.onClearView(recyclerView, viewHolder);
    }

    public interface OnItemTouchCallbackListener<T extends RecyclerView.Adapter> {
        void onSwiped(int position);

        boolean onMove(int beforePosition, int targetPosition, T adapter);
    }

    public interface ItemStatusChange {
        void onSelectChange(RecyclerView.ViewHolder viewHolder, int actionState);
        void onClearView(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder);
    }
}

File: app/src/main/java/com/shen/somebase/widget/manager_drag_adapter_kt/BaseItemTouchHelpCallback.kt
package com.shen.somebase.widget.manager_drag_adapter_kt

import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.RecyclerView

/**
 * created by shen on 2019/10/13
 * at 17:03
 **/
class BaseItemTouchHelpCallback<T : RecyclerView.Adapter<*>>(private val helper: ITouchHelper<T>) : ItemTouchHelper.Callback(){

    override fun getMovementFlags(
        recyclerView: RecyclerView,
        viewHolder: RecyclerView.ViewHolder
    ): Int = if (recyclerView.layoutManager is GridLayoutManager) {
        val dragFlags = ItemTouchHelper.UP or ItemTouchHelper.DOWN or ItemTouchHelper.LEFT or ItemTouchHelper.RIGHT
        val swipeFlags = 0
        makeMovementFlags(dragFlags, swipeFlags)
    } else {
        val dragFlags = ItemTouchHelper.UP or ItemTouchHelper.DOWN
        val swipeFlags = 0
        makeMovementFlags(dragFlags, swipeFlags)
    }

    override fun isLongPressDragEnabled(): Boolean = helper.isLongDragEnable

    override fun isItemViewSwipeEnabled(): Boolean = helper.isSwipeEnable

    @Suppress("UNCHECKED_CAST")
    override fun onMove(
        recyclerView: RecyclerView,
        viewHolder: RecyclerView.ViewHolder,
        target: RecyclerView.ViewHolder
    ): Boolean = helper.onMove?.invoke(viewHolder.adapterPosition, target.adapterPosition, recyclerView.adapter as T) ?: false

    override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
        helper.onSwiped?.invoke(viewHolder.adapterPosition, direction)
    }

    override fun onSelectedChanged(viewHolder: RecyclerView.ViewHolder?, actionState: Int) {
        helper.onSelectChange?.invoke(viewHolder, actionState)
    }

    override fun clearView(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder) {
        helper.onClearView?.invoke(recyclerView, viewHolder)
    }
}

typealias OnSwiped = (position: Int, direction: Int) -> Unit
typealias OnMove = (fromPosition: Int, targetPosition: Int, adapter: RecyclerView.Adapter<*>) -> Boolean
typealias OnSelectChange = (viewHolder: RecyclerView.ViewHolder?, actionState: Int) -> Unit
typealias OnClearView = (recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder) -> Unit
